src/models/StyledComponent.js

```diff
@@ -50,20 +50,32 @@ export default (ComponentStyle: Function) => {
             this.setState({ theme, generatedClassName })
           })
         } else {
+          const theme = this.props.theme || {}
           const generatedClassName = this.generateAndInjectStyles(
-            this.props.theme || {},
+            theme,
             this.props
           )
-          this.setState({ generatedClassName })
+          this.setState({ theme, generatedClassName })
         }
       }
 
       componentWillReceiveProps(nextProps: any) {
+        if (!nextProps.theme) {
+          return
+        }
+
+        // compare if there is any change
+        if (JSON.stringify(nextProps.theme) === JSON.stringify(this.state.theme)) {
```

Comment:
Since the main difference between a deep equality check and the `JSON.stringify` method (which I find very smart btw, have never heard of that before) is that the deep equality check doesn't care about the ordering of the elements. ([see here for more info](http://www.mattzeunert.com/2016/01/28/javascript-deep-equal.html))

How big is the a dependency that can do deep equality?